### PR TITLE
Allow composition of SoftAssertions

### PR DESCRIPTION
**Pull request by ktoublanc**
This pull request add the `assertAlso` method on SoftAssertions class to compose several instances of SoftAssertions.

PS: This is my first pull request, please tell me if you need more information.

#### Check List:
* Unit tests : YES
* Javadoc with a code example (API only) : YES

### CODE REVIEW

**Comment by joel-costigliola** [comment]
@ktoublanc  thanks for the PR, I'd like to ask you what problem it solves that the current soft assertions don't ?
I understand the idea, I just would like to see a real life use case ;-)

**Comment by ktoublanc** [comment]
@joel-costigliola nice question 😄 

I use assert-j on custom testing framework used to test batch processing. 
My use case here is that I am asserting a very complex file content (banking...)
Of course this file is supposed to be following a generic syntax but in real life every client has its custom implementation. 

To be able to test that, my test code is structured in 'strategies' that will assert the file:
- against the generic implementation
- against the client specific implementation

At the end, each specific classes returns a SoftAssertion implementation and I group them to have a consistent error message to help developers understand what is wrong when the test fails.

**Comment by ktoublanc** [comment]
@joel-costigliola do you know if this PR can be merged or if you don't want to have this in assert-j ?

**Comment by joel-costigliola** [comment]
I need a bit more time thinking about it, I'll give you some feedback in the next few days.

**Comment by ktoublanc** [comment]
Ok no problem. Thanks 

Envoyé de mon iPhone

File: src/test/java/org/assertj/core/api/SoftAssertionsTest.java [hunk 1]
```diff
@@ -158,6 +158,34 @@ public void all_assertions_should_pass() {
...
   }
 
+  @Test
+  public void all_composed_assertions_should_pass() {
+    SoftAssertions delegate = new SoftAssertions();
+    delegate.assertThat(1).isEqualTo(1);
+    delegate.assertAll();
+
+    softly.assertThat(1).isEqualTo(1);
+    softly.assertThat(Lists.newArrayList(1, 2)).containsOnly(1, 2);
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/SoftAssertionsTest.java]: Minor: use AssertJ `org.assertj.core.util.List.list` and static import for readability.

**Comment by ktoublanc** [comment]
Cool! Thanks a lot.

I’ll make the changes ASAP.

File: src/test/java/org/assertj/core/api/SoftAssertionsTest.java (hunk 1 above)
Comment (ktoublanc) [reply src/test/java/org/assertj/core/api/SoftAssertionsTest.java]: Done. 
In this case we also could have used Arrays.asList()

File: src/test/java/org/assertj/core/api/SoftAssertionsTest.java [hunk 2]
```diff
@@ -158,6 +158,34 @@ public void all_assertions_should_pass() {
...
+
+    softly.assertThat(1).isEqualTo(2);
+    softly.assertThat(Lists.newArrayList(1, 2)).containsOnly(1, 4);
+    softly.assertAlso(delegate);
+    assertThat(softly.wasSuccess()).isFalse();
+
+    List<Throwable> errorsCollected = softly.errorsCollected();
+    assertThat(errorsCollected).hasSize(2);
+  }
+
```
[the remark this answers is not in the document]
Comment (ktoublanc) [reply src/test/java/org/assertj/core/api/SoftAssertionsTest.java]: Nice catch, my test implementation was actually wrong... 
I fixed that and added GIVEN/WHEN/THEN comments.

File: src/test/java/org/assertj/core/api/SoftAssertionsTest.java [hunk 3]
```diff
@@ -158,6 +158,34 @@ public void all_assertions_should_pass() {
...
+    softly.assertThat(1).isEqualTo(1);
+    softly.assertThat(Lists.newArrayList(1, 2)).containsOnly(1, 2);
+    softly.assertAlso(delegate);
+    softly.assertAll();
+    assertThat(softly.wasSuccess()).isTrue();
+  }
+
+  @Test
+  public void should_work_with_composed_assertions() {
```
[the remark this answers is not in the document]
Comment (ktoublanc) [reply src/test/java/org/assertj/core/api/SoftAssertionsTest.java]: I replaced with should_return_failure_for_failed_composed_assertions
Tell me if you have better naming idea.

File: src/test/java/org/assertj/core/api/SoftAssertionsTest.java [hunk 4]
```diff
@@ -158,6 +158,34 @@ public void all_assertions_should_pass() {
     softly.assertAll();
   }
 
+  @Test
+  public void all_composed_assertions_should_pass() {
```
[the remark this answers is not in the document]
Comment (ktoublanc) [reply src/test/java/org/assertj/core/api/SoftAssertionsTest.java]: Done.

File: src/main/java/org/assertj/core/api/SoftAssertions.java [hunk 5]
```diff
@@ -97,6 +97,37 @@
...
+ *   softly.assertThat(mansion.professor()).as(&quot;Professor&quot;).isEqualTo(&quot;well kempt&quot;);
+ *
+ *   SoftAssertions kitchen = check_kitchen();
+ *   softly.assertAlso(kitchen);
+ *
+ *   SoftAssertions library = check_library();
+ *   softly.assertAlso(library);
+ *
+ *   softly.assertAll();
+ *   });
```
[the remark this answers is not in the document]
Comment (ktoublanc) [reply src/main/java/org/assertj/core/api/SoftAssertions.java]: Nice catch ! 
Done.

**Comment by ktoublanc** [comment]
Hello @joel-costigliola.
Thanks for the review, I amended my commit for git history readability.

**Comment by joel-costigliola** [comment]
Integrated thanks @ktoublanc !

**Comment by ktoublanc** [comment]
Awesome thanks.

Do you know when you plan to release a new version of assertj ?

**Comment by joel-costigliola** [comment]
end of the month or early next month

**Comment by ktoublanc** [comment]
Hello

How are you?

You told me you were thinking releasing assert j last month but I can’t find any new release.

Do you have a new release date in mind?

**Comment by joel-costigliola** [comment]
this week end hopefully

**Pull request closed without merging**